File: app/src/main/java/org/exallium/tradetracker/app/model/Observables.java
package org.exallium.tradetracker.app.model;

import android.net.Uri;
import com.orm.query.Condition;
import com.orm.query.Select;
import org.exallium.tradetracker.app.MainApplication;
import org.exallium.tradetracker.app.R;
import org.exallium.tradetracker.app.model.entities.Card;
import org.exallium.tradetracker.app.model.entities.CardSet;
import org.exallium.tradetracker.app.model.entities.LineItem;
import org.exallium.tradetracker.app.model.entities.Trade;
import org.exallium.tradetracker.app.view.models.CardSetViewModel;
import org.exallium.tradetracker.app.view.models.CardViewModel;
import org.exallium.tradetracker.app.view.models.TradeViewModel;
import org.joda.time.LocalDate;
import rx.Observable;

import java.util.Iterator;
import java.util.List;

public abstract class Observables {

    public static Observable<CardSetViewModel> getCardSetObservable() { return cardSetObservable; }
    public static Observable<TradeViewModel> getTradeObservable() { return tradeObservable; }

    public static Observable<CardViewModel> getCardObservable(String cardSetCode) {
        return Observable.create(subscriber -> {

            final CardSet cardSet = Select.from(CardSet.class).where(Condition.prop("code").eq(cardSetCode)).first();
            final List<Card> cards = Select.from(Card.class)
                    .where(Condition.prop("card_set").eq(cardSet.getId()))
                    .orderBy("name").list();

            for (Card card : cards) {
                subscriber.onNext(new CardViewModel(card.name, cardSetCode));
            }

            subscriber.onCompleted();
        });

    }

    private static Observable<CardSetViewModel> cardSetObservable = Observable.create(subscriber -> {

        final List<CardSet> cardSets = Select.from(CardSet.class).orderBy("name").list();
        for (CardSet cardSet : cardSets) {
            subscriber.onNext(new CardSetViewModel(cardSet.code, cardSet.name));
        }

        subscriber.onCompleted();
    });

    private static Observable<TradeViewModel> tradeObservable = Observable.create(subscriber -> {

        final List<Trade> trades = Select.from(Trade.class).orderBy("trade_date").list();
        for (Trade trade : trades)
            subscriber.onNext(trade);

        subscriber.onCompleted();

    }).map(t -> {

        Trade trade = (Trade) t;

        List<LineItem> lineItems = Select.from(LineItem.class).where(Condition.prop("trade").eq(trade.getId())).list();

        long value = lineItems.isEmpty() ? 0 : Observable.from(lineItems)
                .map(lineItem -> lineItem.value)
                .reduce((l1, l2) -> l1 + l2)
                .map(l1 -> l1 / 100)
                .toBlocking().last();

        Iterator<Card> cardIterator = Observable.from(lineItems)
                .map(lineItem -> lineItem.card)
                .filter(c -> c != null)
                .distinct(card1 -> card1.name)
                .toBlocking().getIterator();

        Iterator<String> descIterator = Observable.from(lineItems)
                .map(lineItem -> lineItem.description)
                .filter(d -> d != null)
                .distinct()
                .toBlocking().getIterator();

        Uri imageUri = null;
        StringBuilder itemsTradedBuilder = new StringBuilder();

        while (cardIterator.hasNext()) {
            Card card = cardIterator.next();
            if (imageUri == null && card.multiverseId != -1)
                imageUri = Uri.parse("http://gatherer.wizards.com/handlers/image.ashx/?type=card&multiverseid=" + card.multiverseId);
            itemsTradedBuilder.append(String.format("%s [%s], ", card.name, card.cardSet.code));
        }

        while (descIterator.hasNext()) {
            String description = descIterator.next();
            itemsTradedBuilder.append(String.format("%s, ", description));
        }

        if (itemsTradedBuilder.length() > 0)
            itemsTradedBuilder.deleteCharAt(itemsTradedBuilder.length() - 2);
        else
            itemsTradedBuilder.append(MainApplication.getInstance().getResources().getString(R.string.no_line_items_available));

        return new TradeViewModel(
                trade.getId(),
                String.format("%c%d", value >= 0 ? '+' : '-', value),
                trade.person.name,
                imageUri,
                lineItems.size(),
                itemsTradedBuilder.toString(),
                LocalDate.fromDateFields(trade.tradeDate)
        );
    });

}

File: app/src/main/kotlin/org/exallium/tradetracker/app/controller/observables.kt
package org.exallium.tradetracker.app.model

import android.net.Uri
import com.orm.query.Condition
import com.orm.query.Select
import org.exallium.tradetracker.app.R
import org.exallium.tradetracker.app.controller.BundleConstants
import org.exallium.tradetracker.app.controller.MainApplication
import org.exallium.tradetracker.app.model.entities.*
import org.exallium.tradetracker.app.view.models.*
import org.joda.time.LocalDate
import rx.Observable

public object Observables {

    public fun getLineItemObservable(direction: Boolean, tradeId: Long): Observable<LineItemViewModel> {
        return Observable.create<LineItem> { subscriber ->
            if (tradeId != BundleConstants.NEW_OBJECT) {
                Select.from(javaClass<LineItem>())
                        .where(Condition.prop("trade").eq(tradeId), Condition.prop("direction").eq(if (direction) 1 else 0))
                        .list().forEach { lineItem ->
                    subscriber.onNext(lineItem)
                }
            }
            subscriber.onCompleted()
        }.map { lineItem ->
            val description = if (lineItem.card != null) "%s [%s]".format(lineItem.card?.name, lineItem.card?.cardSet?.code) else lineItem.description
            LineItemViewModel(description, lineItem.quantity)
        }
    }

    public fun getCardObservable(cardSetCode: String): Observable<CardViewModel> {
        var cardSet = Select.from(javaClass<CardSet>()).where(Condition.prop("code").eq(cardSetCode)).first()
        return Observable.from(Select.from(javaClass<Card>()).where(Condition.prop("card_set").eq(cardSet.getId())).list())
                .map { card -> CardViewModel(card.name, cardSet.code) }
    }

    public fun getCardSetObservable(): Observable<CardSetViewModel> {
        return Observable.create { subscriber ->
            val cardSets = Select.from(javaClass<CardSet>()).orderBy("name").list()
            cardSets.forEach { cardSet ->
                subscriber.onNext(CardSetViewModel(cardSet.code, cardSet.name))
            }
            subscriber.onCompleted()
        }
    }

    public fun getPersonObservable(): Observable<PersonViewModel> {
        return Observable.from(Select.from(javaClass<Person>()).list()).map {
            person -> PersonViewModel(person.name)
        }
    }

    public fun getTradeObservable(): Observable<TradeViewModel> {

        return Observable.from(
                Select.from(javaClass<Trade>()).where(Condition.prop("is_temporary").eq(0)).orderBy("trade_date").list().reverse()
        ).map { trade ->
            val lineItems = Select.from(javaClass<LineItem>()).where(Condition.prop("trade").eq(trade.getId())).list()
            val tradeValue = if (lineItems.size() == 0) 0 else Observable.from(lineItems)
                    .map { item -> item.value * item.quantity * (if (item.direction) 1 else -1) }
                    .reduce { item1, item2 -> item1 + item2 }
                    .map { item -> item / 100L }
                    .toBlocking().last()
            val cardIterator = Observable.from(lineItems)
                    .map { lineItem -> lineItem.card }
                    .filter { c -> c != null }
                    .distinct { card1 -> card1?.name }
                    .toBlocking().getIterator()
            val descIterator = Observable.from<LineItem>(lineItems)
                    .map<String> { lineItem -> lineItem.description }
                    .filter { d -> d != null }
                    .distinct().toBlocking().getIterator()

            var imageUri: Uri? = null
            val itemsTradedBuilder = StringBuilder()

            for (card in cardIterator) {
                if (imageUri == null && card?.multiverseId != -1)
                    imageUri = Uri.parse("http://gatherer.wizards.com/handlers/image.ashx/?type=card&multiverseid=" + card?.multiverseId);
                itemsTradedBuilder.append("%s [%s], ".format(card?.name, card?.cardSet?.code))
            }

            for (desc in descIterator) {
                itemsTradedBuilder.append("%s, ".format(desc))
            }

            if (itemsTradedBuilder.length() > 0)
                itemsTradedBuilder.deleteCharAt(itemsTradedBuilder.length() - 2)
            else
                itemsTradedBuilder.append(MainApplication.getResources()?.getString(R.string.no_line_items_available))

            TradeViewModel(
                    trade.getId(),
                    "%c%d".format(if (tradeValue >= 0) '+' else 0, tradeValue),
                    trade.person?.name,
                    imageUri,
                    lineItems.size(),
                    itemsTradedBuilder.toString(),
                    LocalDate.fromDateFields(trade.tradeDate)
            )
        }
    }
}





